common/util/trees.h

```diff
@@ -0,0 +1,856 @@
+  return RightmostDescendant(*prev_ancestor);
+}
+
+// Returns the next sibling node if it exists, else nullptr.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::Parent::available>* = nullptr>
+T* NextSibling(T& node) {
+  if (node.Parent() == nullptr) {
+    return nullptr;
+  }
+  const size_t birth_rank = BirthRank(node);
+  const size_t next_rank = birth_rank + 1;
+  if (next_rank == node.Parent()->Children().size()) {
+    return nullptr;  // This is the last child of the Parent().
+  }
+  // More children follow this one.
+  return &*std::next(node.Parent()->Children().begin(), next_rank);
+}
+
+// Returns the previous sibling node if it exists, else nullptr.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::Parent::available>* = nullptr>
+T* PreviousSibling(T& node) {
+  if (node.Parent() == nullptr) {
+    return nullptr;
+  }
+  const size_t birth_rank = BirthRank(node);
+  if (birth_rank == 0) {
+    return nullptr;
+  }
+  // More children precede this one.
+  return &*std::next(node.Parent()->Children().begin(), birth_rank - 1);
+}
+
+// Removes this node from its parent, and shifts ths siblings that follow this
+// node lower in BirthRank().
+// Any iterators that pointed to this node or its later siblings are
+// invalidated.
+// This node is destroyed in the process.
+// This operation is only valid on non-root nodes.
+// It is the caller's responsibility to maintain invariants before
+// destroying this node.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::Parent::available &&
+                           !std::is_const_v<T>>* = nullptr>
+void RemoveSelfFromParent(T& node) {
+  auto& siblings = ABSL_DIE_IF_NULL(node.Parent())->Children();
+  auto self_iter = std::next(siblings.begin(), verible::BirthRank(node));
+  CHECK_EQ(&*self_iter, &node);
+  siblings.erase(self_iter);
+}
+
+// Recursively transform a SrcTree to DstTree.
+// This cannot be expressed as a constructor because template constructor
+// calls must be type-deducible from its constructor args, and there's no
+// way to explicitly invoke a constructor with specific template args.
+// The resulting tree is always StructureEqual() to the original tree.
+//
+// Usage:
+//   auto tree_other = orig_tree.Transform<OtherType>(
+//       [](const SrcTree& node) { return ... });
+template <
+    class DstTree, class SrcTree, class SrcNodeToDstValueFunc,
+    class DstValue_ =
+        std::invoke_result_t<SrcNodeToDstValueFunc, const SrcTree&>,  //
+    std::enable_if_t<TreeNodeTraits<DstTree>::Value::available &&
+                     TreeNodeTraits<SrcTree>::Value::available &&
+                     std::is_constructible_v<DstTree, DstValue_>>* = nullptr>
+DstTree Transform(const SrcTree& src_node, const SrcNodeToDstValueFunc& f) {
+  // Using invoke() to allow passing SrcTree's method pointers as `f`
+  DstTree dst_node(std::invoke(f, src_node));
+  dst_node.Children().reserve(src_node.Children().size());
```

Comment:
Worth noting that this implementation requires the children container to support `reserve`.  This was done to guarantee node address-stability.

(Probably worth commenting in other functions that call `reserve`)

Reply:
I've changed it (and other similar functions) to call `reserve()` only if supported by the container.